### PR TITLE
Add "View details" button in Options menu

### PR DESCRIPTION
**Pull request by CenTdemeern1**
See also: https://github.com/ppy/osu/discussions/21249#discussioncomment-4243822

This PR adds a "View details" button in the options menu in the song select, allowing for opening the selected beatmap's listing on mobile devices.

osu_2022-12-02_11-48-59

Apparently, nothing in this menu has tests? It might be good to add tests for everything in a separate PR. If it's needed, that is.

### CODE REVIEW

**Comment by peppy** [comment]
At 4:3 this no longer fits on the screen:

2022-12-03 22 14 32@2x

Not sure if this is cause for concern or not (depends on how these options will exist post-redesign).

**Comment by CenTdemeern1** [comment]
> The claim that ["there is no method for inserting a button at a specified position"](https://github.com/ppy/osu/discussions/21249#discussioncomment-4291814) doesn't really hold because introducing one is relatively trivial. See `FlowContainer.SetLayoutPosition()`.

I'll look into that.

**Comment by CenTdemeern1** [comment]
> At 4:3 this no longer fits on the screen:
> 
> [image]
> 
> Not sure if this is cause for concern or not (depends on how these options will exist post-redesign).

Suggested solution: I could maybe center the contents of the options menu on the screen? I thought it was already centered, guess not

**Comment by CenTdemeern1** [comment]
Apologies for the week of silence, but I have made it work

Play song select
osu_2022-12-11_19-27-28

Multiplayer song select
osu_2022-12-11_19-27-18

**Comment by CenTdemeern1** [comment]
At least, only with the layout. The keyboard shortcut is not affected by `SetLayoutPosition` so I am going to have to do something about that
I will try to use `Insert` instead of `SetLayoutPosition` to see if that works

**Comment by CenTdemeern1** [comment]
@bdach do you have any ideas in regards to making this work with the keyboard shortcuts?

File: osu.Game/Screens/Select/SongSelect.cs [hunk 1]
```diff
@@ -290,6 +293,7 @@ private void load(AudioManager audio, OsuColour colours, ManageCollectionsDialog
                 BeatmapOptions.AddButton(@"Delete", @"all difficulties", FontAwesome.Solid.Trash, colours.Pink, () => delete(Beatmap.Value.BeatmapSetInfo));
                 BeatmapOptions.AddButton(@"Remove", @"from unplayed", FontAwesome.Regular.TimesCircle, colours.Purple, null);
                 BeatmapOptions.AddButton(@"Clear", @"local scores", FontAwesome.Solid.Eraser, colours.Purple, () => clearScores(Beatmap.Value.BeatmapInfo));
+                BeatmapOptions.InsertButton(2, @"View", @"details", FontAwesome.Solid.Bars, colours.Blue, () => beatmapOverlay?.FetchAndShowBeatmap(Beatmap.Value.BeatmapInfo.OnlineID));
```
Comment (peppy) [inline osu.Game/Screens/Select/SongSelect.cs:296]: Why is this required? Why can't you just move it up?

Comment (CenTdemeern1) [reply osu.Game/Screens/Select/SongSelect.cs:296]: Move it up in what way?

Comment (frenzibyte) [reply osu.Game/Screens/Select/SongSelect.cs:296]: This is required because inheritors of `SongSelect` also add their own buttons (e.g. `PlaySongSelect` adding "Edit Beatmap"), but this button wants to be ordered last, so it uses a depth of its own that makes it placed last.

This can work alternative by just adding a virtual method for inheritors to override and add buttons, then place the view details button afterwards, i.e.:
```csharp
                ...
                BeatmapOptions.AddButton(@"Remove", @"from unplayed", FontAwesome.Regular.TimesCircle, colours.Purple, null);
                BeatmapOptions.AddButton(@"Clear", @"local scores", FontAwesome.Solid.Eraser, colours.Purple, () => clearScores(Beatmap.Value.BeatmapInfo));
                
                AddButtons();
                
                BeatmapOptions.AddButton(@"View", @"details", FontAwesome.Solid.Bars, colours.Blue, () => beatmapOverlay?.FetchAndShowBeatmap(Beatmap.Value.BeatmapInfo.OnlineID));

Comment (CenTdemeern1) [reply osu.Game/Screens/Select/SongSelect.cs:296]: Using a virtual method for this could work, but it would only allow buttons to be inserted into a specific location instead of an arbitrary one. It could definitely be enough though, and it's definitely another option

Comment (CenTdemeern1) [reply osu.Game/Screens/Select/SongSelect.cs:296]: Does "the proposed solution" refer to the virtual method? If so I'll implement that asap.

Comment (peppy) [reply osu.Game/Screens/Select/SongSelect.cs:296]: Yes.

**Comment by CenTdemeern1** [comment]
Wait- I didn’t click that button-

**Comment by CenTdemeern1** [comment]
Right… that didn’t work
To resolve the merge conflicts I am actually going to have to merge the master branch into my branch, which is less than ideal

**Comment by peppy** [comment]
> To resolve the merge conflicts I am actually going to have to merge the master branch into my branch, which is less than ideal

That's literally the only way to resolve conflicts, since the conflicts are between `master` and your branch. I've fixed the merge for you, although you'll definitely want to look into the fact that you're sending a `protected` field from `SongSelect` as a parameter in `AddOptionButtons` for probably no reason.

**Comment by CenTdemeern1** [comment]
Oh yeah whoops that's a mistake, I was kind of tired when I did that. Thanks for the catch, and apologies for the mistake.
Regarding the conflict, I recently contributed to another repo that wants to avoid merging master into itself at all costs, so that has kind of twisted my view on this. Again, apologies.

**Comment by peppy** [comment]
Probably best to add this functionality only to the V2 footer? See #21491.

**Comment by Joehuu** [comment]
I believe https://github.com/ppy/osu/pull/21497#issuecomment-1341898341 is the better solution. The new options popover now has headers and a hide button so adding more clutters it more.

**Comment by peppy** [comment]
> I believe [#21497 (comment)](https://github.com/ppy/osu/pull/21497#issuecomment-1341898341) is the better solution. The new options popover now has headers and a hide button so adding more clutters it more.

Not 100% sure about this, there's still room for a few more options, but we'll see after the new design comes together I guess.
